### PR TITLE
feat(search): add onClear callback prop to Search/DataTableSearch

### LINKED ISSUE #9738: [Search]: add onClear callback when input is cleared.
### Summary

We need a way in the carbon-addons-iot-react repo to only trigger search when a user hits Enter or blurs the input. The problem is with all changes pushed through onChange there's no way to tell if they backspaced to delete everything or if the "Clear search input" button was clicked.

### Justification

To improve performance in complex tables by only trigger server calls when the user specifically hits enter or blurs the input. The onClear callback allows us to know the difference between multiple backspaces and a clear. For example, they may want to backspace and delete everything, type a new search and then hit 'Enter' to trigger it.

### Desired UX and success metrics

_No response_

### Required functionality

_No response_

### Specific timeline issues / requests

_No response_

### Available extra resources

I can write the PR.

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/carbon-design-system/carbon/blob/f555616971a03fd454c0f4daea184adf41fff05b/.github/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by kevinsperrine**
Closes #9738

We needed a way in the `carbon-addons-iot-react` repo to only trigger search when a user hits `Enter` or blurs the input. The problem was with all changes pushed through `onChange` there's no way to tell if they backspaced to delete everything or if the "Clear search input" button was clicked. This PR adds an `onClear` callback prop to distinguish between the two.

#### Changelog

**New**

- added onClear to the Search/DataTableSearch components

#### Testing / Reviewing

Added new tests to Search component to confirm callback is called.

### CODE REVIEW

**Review by tay1orjones — COMMENTED** [review]
Thanks for the PR @kevinsperrine ! 🎉  

This addition makes sense to me. Another option would be to modify the signature of the `onChange` to include a new `isClear` param or something, but based on https://github.com/carbon-design-system/carbon/issues/9641 I don't think we'd want to do that.

I see the `onClear` fire in the DataTable storybook, but I don't see it firing in the Search story?

https://user-images.githubusercontent.com/3360588/134743664-993ddcd3-90ec-4570-aad6-c91d8f0bb3fd.mov

Otherwise this looks solid 👍

**Comment by kevinsperrine** [comment]
> I see the `onClear` fire in the DataTable storybook, but I don't see it firing in the Search story?

@tay1orjones Thanks for the heads up. I have no idea what's going on in the deploy preview. Locally, there are onChange, onKeyDown, and the onClear handlers being fired, but only onChange in the preview. 🤔 

search-actions

**Review by tay1orjones — APPROVED** [review]
@kevinsperrine I'm not sure what's going on there. Locally I see the same though; `onKeyDown` and `onClear` logging as expected to the actions pane. It might be something with the deploy preview environment and the fact that you're a first time contributor so some status checks had to be approved to be ran. The production storybook was/is logging things fine. 

I'm good to merge this, if the storybook issue persists in production I can open a new issue to hunt it down.

**Review by abbeyhrt — APPROVED** [review]
LGTM! Thanks for the contribtion!

File: packages/react/src/components/Search/Search-test.js [hunk 1]
```diff
@@ -208,6 +210,12 @@ describe('Search', () => {
         input.simulate('change', eventObject);
         expect(onChange).toHaveBeenCalledWith(eventObject);
       });
+
+      it('should invoke onClear when input value is cleared', () => {
+        wrapper.setProps({ defaultValue: 'test' });
+        wrapper.find('button').simulate('click');
+        expect(onClear).toHaveBeenCalled();
+      });
```
Comment (jnm2377) [inline packages/react/src/components/Search/Search-test.js]: it looks like this test is failing

Screen Shot 2021-10-06 at 1 39 14 AM

Comment (jnm2377) [reply packages/react/src/components/Search/Search-test.js]: @kevinsperrine we can merge once this is updated and tests pass :)

Comment (kevinsperrine) [reply packages/react/src/components/Search/Search-test.js]: @jnm2377 Thanks. It's fixed now. I adjusted the test to account for shallow rendering and all is well.

**Pull request merged**